src/gui/tagfiltermodel.cpp

```diff
@@ -0,0 +1,326 @@
+    const QModelIndex i = index(item);
+    emit dataChanged(i, i);
+}
+
+void TagFilterModel::torrentTagRemoved(const QString &tag, int new_tag_count)
+{
+    if (new_tag_count <= 0)
+        findItem(getSpecialUntaggedTag())->increaseTorrentsCount();
+
+    TagModelItem *item = findItem(tag);
+    Q_ASSERT(item);
+
+    item->decreaseTorrentsCount();
+    const QModelIndex i = index(item);
+    emit dataChanged(i, i);
+}
+
+void TagFilterModel::torrentAdded(BitTorrent::TorrentHandle *const torrent)
+{
+    findItem(getSpecialAllTag())->increaseTorrentsCount();
+
+    const QVector<TagModelItem *> items = findItems(torrent->tags());
+    if (items.isEmpty())
+        findItem(getSpecialUntaggedTag())->increaseTorrentsCount();
+
+    foreach (TagModelItem *item, items)
+        item->increaseTorrentsCount();
+}
+
+void TagFilterModel::torrentAboutToBeRemoved(BitTorrent::TorrentHandle *const torrent)
+{
+    findItem(getSpecialAllTag())->decreaseTorrentsCount();
+
+    if (torrent->tags().isEmpty())
+        findItem(getSpecialUntaggedTag())->decreaseTorrentsCount();
+
+    foreach (TagModelItem *item, findItems(torrent->tags()))
+        item->decreaseTorrentsCount();
+}
+
+const QString &TagFilterModel::getSpecialAllTag()
+{
+    static const QString *const ALL_TAG = new QString(" ");
+    Q_ASSERT(!BitTorrent::Session::isValidTag(*ALL_TAG));
+    return *ALL_TAG;
+}
+
+const QString &TagFilterModel::getSpecialUntaggedTag()
+{
+    static const QString *const UNTAGGED_TAG = new QString("  ");
+    Q_ASSERT(!BitTorrent::Session::isValidTag(*UNTAGGED_TAG));
+    return *UNTAGGED_TAG;
+}
+
+void TagFilterModel::populate()
+{
+    using Torrent = BitTorrent::TorrentHandle;
+
+    auto session = BitTorrent::Session::instance();
+    auto torrents = session->torrents();
+
+    // All torrents
+    addItemToModel(new TagModelItem(getSpecialAllTag(), tr("All"), torrents.count()));
+
+    const int untagged_count = std::count_if(torrents.begin(), torrents.end(),
+                                             [] (Torrent * torrent) { return torrent->tags().isEmpty(); });
+    addItemToModel(new TagModelItem(getSpecialUntaggedTag(), tr("Untagged"), untagged_count));
+
+    foreach (const QString &tag, session->tags()) {
+        const int count = std::count_if(torrents.begin(), torrents.end(),
+                                        [tag](Torrent * torrent) { return torrent->hasTag(tag); });
+        addItemToModel(new TagModelItem(tag, tag, count));
+    }
+}
+
+void TagFilterModel::addItemToModel(TagModelItem *item)
+{
+    Q_ASSERT(item);
+    const QString &tag = item->tag();
+    Q_ASSERT(!m_tag_rows.contains(tag));
+    m_tag_rows.append(tag);
+    m_tag_items[tag] = item;
+}
+
+void TagFilterModel::removeItemFromModel(TagModelItem *item)
+{
+    const QString &tag = item->tag();
+    m_tag_rows.removeOne(tag);
+    m_tag_items.remove(tag);
+    delete item;
+}
+
+TagModelItem *TagFilterModel::findItem(const QString &tag) const
```

Comment:
It's probably worth keeping as a convenience wrapper since it gets called a lot, but I used QMap::value to make it a little more succinct.